ComponentKit/StatefulViews/CKStatefulViewReusePool.mm

```diff
@@ -10,40 +10,74 @@
 #import "CKStatefulViewComponentController.h"
 
 #import "CKStatefulViewReusePool.h"
+#import "CKAssert.h"
 
 #import <unordered_map>
 
+struct FBStatefulReusePoolItemEntry {
+  UIView *view;
+  CKStatefulViewReusePoolPendingMayRelinquishBlock block;
+};
+
 class FBStatefulReusePoolItem {
 public:
-  FBStatefulReusePoolItem()
-  : views([NSMutableArray array]) {};
-
   UIView *viewWithPreferredSuperview(UIView *preferredSuperview)
   {
-    if ([views count] == 0) {
+    if (_entries.empty()) {
       return nil;
     }
-    const NSUInteger matchingIndex = [views indexOfObjectPassingTest:^BOOL(UIView *view, NSUInteger idx, BOOL *stop) {
-      return [view superview] == preferredSuperview;
-    }];
-    const NSUInteger viewIndex = (matchingIndex == NSNotFound) ? 0 : matchingIndex;
-    UIView *view = views[viewIndex];
-    [views removeObjectAtIndex:viewIndex];
-    return view;
+    // Preferentially return the parent view.
+    auto preferIt = std::find_if(_entries.begin(), _entries.end(),
+                           [preferredSuperview](const FBStatefulReusePoolItemEntry entry)->bool {
+                             return entry.view == preferredSuperview;
+                           });
+    if (preferIt != _entries.end()) {
+      FBStatefulReusePoolItemEntry entry = *preferIt;
+      _entries.erase(preferIt);
+      if (entry.block == NULL || entry.block()) {
+        return entry.view;
+      }
+    }
+
+    // We didn't find the item preferentially. Time to fall back to going from start to finish.
+    auto it = _entries.begin();
+    while (it != _entries.end()) {
+      FBStatefulReusePoolItemEntry entry = *it;
+      // erase returns the next iterator
+      it = _entries.erase(it);
+      if (entry.block == NULL || entry.block()) {
+        // The block tells us it's OK to reuse this view
+        return entry.view;
+      }
+    }
+
+    return nil;
   };
   
   NSUInteger viewCount()
   {
-    return [views count];
+    return _entries.size();
   };
 
-  void addView(UIView *view)
+  void addEntry(const FBStatefulReusePoolItemEntry &entry)
   {
-    [views addObject:view];
+    _entries.push_back(entry);
   };
 
+  void absorbPendingPool(const FBStatefulReusePoolItem &otherPool, NSInteger maxEntries)
+  {
+    for (const FBStatefulReusePoolItemEntry &entry : otherPool._entries) {
+      if (entry.block == NULL || entry.block()) {
+        // The block was either not present, or evaluated to true, indicating it can join the normal reuse pool
```

Comment:
I'd rephrase this comment to mention why the entry can join the normal reuse pool. The logic is simple enough to follow so I don't see much of a need to reiterate it here. Why that condition is important is a bit more interesting.